src/services/factories/getPriceEstimation.ts

```diff
@@ -0,0 +1,73 @@
+import BigNumber from 'bignumber.js'
+import { assert } from '@gnosis.pm/dex-js'
+
+import { ZERO_BIG_NUMBER } from 'const'
+
+import { TheGraphApi } from 'api/thegraph/TheGraphApi'
+import { TokenList } from 'api/tokenList/TokenListApi'
+import { Network } from 'types'
+
+export interface GetPriceParams {
+  numeratorTokenId: number
+  denominatorTokenId: number
+}
+
+export function getPriceEstimationFactory(factoryParams: {
+  theGraphApi: TheGraphApi
+  tokenListApi: TokenList
+}): (params: GetPriceParams) => Promise<BigNumber> {
+  const { theGraphApi, tokenListApi } = factoryParams
+
+  // Only make sense to fetch prices for mainnet, thus we won't accept networkId parameter on this service
+  const networkId = Network.Mainnet
+
+  const tokenIdsSet = new Set(tokenListApi.getTokens(networkId).map(token => token.id))
+  const tokenIds = Array.from(tokenIdsSet).sort()
+
+  /**
+   * Given a tokenId, checked whether it's already in our list.
+   * Returns true if tokenId was added to the list and it requires sorting
+   */
+  function addTokenId(tokenId: number): boolean {
+    if (!tokenIdsSet.has(tokenId)) {
+      tokenIdsSet.add(tokenId)
+
+      tokenIds.push(tokenId)
+
+      return true
+    }
+
+    return false
+  }
+
+  return async (params: GetPriceParams): Promise<BigNumber> => {
+    const { numeratorTokenId, denominatorTokenId } = params
+
+    assert(numeratorTokenId !== denominatorTokenId, 'Token ids cannot be the equal')
+
+    // As much as it'd be visually pleasing, cannot simplify it to a single expression
+    // We need to execute both functions.
+    let needsToSort = addTokenId(numeratorTokenId)
+    needsToSort = addTokenId(denominatorTokenId) || needsToSort
+
+    if (needsToSort) {
+      tokenIds.sort()
+    }
+
+    const prices = await theGraphApi.getPrices({ networkId, tokenIds })
+
+    if (denominatorTokenId === 0) {
+      // Optimization more for a matter of principle than performance:
+      // OWL token is always tokenId == 0 on the contract
+      // All prices are given in terms of OWL
+      // OWL price is always 1
+      // Anything divided by 1...
+      return prices[numeratorTokenId]
```

Comment:
I see the point, but why doing this optimization, and not the case where numerator is 0, so the result is 1/price

I would remove this optimization and just make the:
```js
numeratorPriceInOwl = numeratorTokenId === 0 ? ONE : prices[numeratorTokenId]
denominatorPriceInOwl = denominatorTokenId === 0 ? ONE : prices[denominatorTokenId]
return numeratorPriceInOwl.dividedBy(denominatorPriceInOwl)
```

Reply:
Done